filter.go

```diff
@@ -43,31 +43,66 @@ const (
 	MapResultSkipDir
 )
 
-func Walk(ctx context.Context, p string, opt *WalkOpt, fn filepath.WalkFunc) error {
-	root, err := filepath.EvalSymlinks(p)
-	if err != nil {
-		return errors.WithStack(&os.PathError{Op: "resolve", Path: root, Err: err})
-	}
-	rootFI, err := os.Stat(root)
-	if err != nil {
-		return errors.WithStack(err)
+type filterFS struct {
+	fs  FS
+	opt *FilterOpt
+}
+
+func NewFilterFS(fs FS, opt *FilterOpt) FS {
+	if opt == nil {
+		return fs
 	}
-	if !rootFI.IsDir() {
-		return errors.WithStack(&os.PathError{Op: "walk", Path: root, Err: syscall.ENOTDIR})
+
+	if fs, ok := fs.(*filterFS); ok {
+		includes := append([]string{}, fs.opt.IncludePatterns...)
+		includes = append(includes, opt.IncludePatterns...)
+		excludes := append([]string{}, fs.opt.ExcludePatterns...)
+		excludes = append(excludes, opt.ExcludePatterns...)
+		follows := append([]string{}, fs.opt.FollowPaths...)
+		follows = append(follows, opt.FollowPaths...)
+		var mapFn MapFunc
+		if fs.opt.Map != nil && opt.Map != nil {
+			mapFn = func(path string, stat *types.Stat) MapResult {
+				result := fs.opt.Map(path, stat)
+				if result != MapResultKeep {
+					return result
+				}
+				return opt.Map(path, stat)
+			}
+		} else if fs.opt.Map != nil {
+			mapFn = fs.opt.Map
+		} else if opt.Map != nil {
+			mapFn = opt.Map
+		}
+		return &filterFS{fs, &FilterOpt{
+			IncludePatterns: includes,
+			ExcludePatterns: excludes,
+			FollowPaths:     follows,
+			Map:             mapFn,
+		}}
 	}
 
+	return &filterFS{fs, opt}
+}
+
+func (fs *filterFS) Open(p string) (io.ReadCloser, error) {
+	return fs.fs.Open(p)
```

Comment:
This isn't really correct iiuc. Eg. when asking for a file that is filtered out it would still be opened.